### ISSUE TITLE
CCurlFile::Open failed with code 404 due to CURL::Decode/Encode mangling URLs with character entities

### ISSUE THREAD

**Issue by wagnerch** [xbmc/xbmc#18413]
### Describe the bug

Using HTTP sources, if a folder contains a character that is encoded to a character entity (i.e. &) then Apache AutoIndex will encode the HREF as:
```"A%20Very%20Harold%20&amp;%20Kumar%20Christmas%20(2011)/"```

It appears that at some point this is re-encoded using CURL::Encode, which translates the part into this:
```"A%20Very%20Harold%20&amp%3b%20Kumar%20Christmas%20(2011)/"```

## Expected Behavior

Kodi should handle URLs encoded with character entities in CHTTPDirectory::GetDirectory, since HTML spec allows character entities in anchor href's (since the data type is a CDATA).

## Actual Behavior

Kodi mangles URLs that contain character entities, and video scanning fails to open the folder.

### To Reproduce

1. Create folder with "&" in name, using movie NFO structure.
2. Start video scan.
3. Review kodi log and you should see CCurlFile::Open failed with code 404 failures

### Debuglog

The debuglog can be found here:
https://paste.kodi.tv/raw/ifacumipun

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Raspberry Pi OS Buster (Debian based)
 - Kodi version: 18.7.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by wagnerch** [comment]
It looks like the issue is [L#134](https://github.com/xbmc/xbmc/blob/75cafc67cdd55646d7d01ac1cb588d66640c6f0f/xbmc/filesystem/HTTPDirectory.cpp#L134) needs to be moved after L#138.  

HTML::CHTMLUtil::ConvertHTMLToW will decode the HTML entities, but since CURL::Decode is called first it is converting the semicolon to %3b.

**Comment by wagnerch** [comment]
It looks like the problem is in CHTTPDirectory::GetDirectory, with [this](https://github.com/xbmc/xbmc/blob/2106a2e75492ce12468217076b2868c55dd0b903/xbmc/filesystem/HTTPDirectory.cpp#L122-L129) section of code.  It seems like perhaps [this](https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/HTTPDirectory.cpp#L134-L139) section should be done first.

**Comment by wagnerch** [comment]
Here is another issue with HTTP directory parsing, not sure what is going on but I am guessing it fails to detect the character set properly and encode/decode character sets.

```
HTTP/1.1 200 OK
Date: Wed, 30 Sep 2020 22:18:46 GMT
Server: Apache/2.4.29 (Ubuntu)
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 1402
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
**Content-Type: text/html;charset=UTF-8**
```

...

```
<a href="Shameless%20(US)%20-%20S05E04%20-%20A%20Night%20to%20Remem%e2%80%a6%20Wait,%20What.mkv">Shameless (US) - S05E04 - A Night to Remem… Wait, What.mkv</a>
```

Here the filename is encoded with ellipses "..." UTF-8 byte string e2 80 a6, which the href is properly encoded and the anchor text is also UTF-8 byte string e2 80 a6.  I know the code is trying to match the anchor text to the filename, so perhaps one of them is not handling UTF-8 properly.
